packages/registry/contracts/CannonSubscription.sol

```diff
@@ -0,0 +1,157 @@
+ */
+contract CannonSubscription is ReentrancyGuard {
+  using Subscription for Subscription.Data;
+
+  error ZeroAddressNotAllowed(string variableName);
+  error InsufficientAllowance(address user, uint256 allowance, uint256 required);
+  error TransferFailed();
+  error PriceOverflow();
+  error MembershipNotActive(address user);
+
+  event PlanRegistered(
+    uint16 indexed planId,
+    uint32 termDuration,
+    uint32 quota,
+    uint16 minTerms,
+    uint16 maxTerms,
+    uint256 price
+  );
+  event PlanSetAsDefault(uint16 indexed planId);
+  event CreditsUsed(address indexed user, uint32 creditsUsed, uint32 remainingCredits);
+  event MembershipPurchased(address indexed user, uint32 amountOfTerms, uint256 totalPrice);
+  event MembershipCancelled(address indexed user, uint32 pendingTerms, uint256 reimbursement);
+
+  IERC20 public immutable USDC;
+  address public immutable VAULT;
+
+  constructor(address _usdcAddress, address _vaultAddress) {
+    if (_usdcAddress == address(0)) revert ZeroAddressNotAllowed("usdcAddress");
+    if (_vaultAddress == address(0)) revert ZeroAddressNotAllowed("vaultAddress");
+
+    USDC = IERC20(_usdcAddress);
+    VAULT = _vaultAddress;
+  }
+
+  function getPlan(uint16 _planId) external view returns (Subscription.Plan memory) {
+    return Subscription.load().getPlan(_planId);
+  }
+
+  function getDefaultPlan() external view returns (Subscription.Plan memory) {
+    return Subscription.load().getDefaultPlan();
+  }
+
+  function registerPlan(
+    uint32 _termDuration,
+    uint32 _quota,
+    uint16 _minTerms,
+    uint16 _maxTerms,
+    uint256 _price
+  ) external returns (uint16) {
+    OwnableStorage.onlyOwner();
+
+    uint16 planId = Subscription.load().registerPlan(_termDuration, _quota, _minTerms, _maxTerms, _price);
+    emit PlanRegistered(planId, _termDuration, _quota, _minTerms, _maxTerms, _price);
+
+    return planId;
+  }
+
+  function setDefaultPlan(uint16 _planId) external {
+    OwnableStorage.onlyOwner();
+    Subscription.load().setDefaultPlan(_planId);
+    emit PlanSetAsDefault(_planId);
+  }
+
+  function updatePlanStatus(uint16 _planId, bool _isActive) external {
+    OwnableStorage.onlyOwner();
+    Subscription.load().updatePlanStatus(_planId, _isActive);
+  }
+
+  function getMembership(address _user) external view returns (Subscription.Membership memory) {
+    return Subscription.load().getMembership(_user);
+  }
+
+  function purchaseMembership(uint32 _amountOfTerms) external nonReentrant {
+    address _sender = ERC2771Context.msgSender();
+    Subscription.Data storage _subscription = Subscription.load();
+
+    Subscription.Membership storage _membership = _subscription.getMembership(_sender);
+    Subscription.Plan storage _plan = _membership.planId == 0
+      ? _subscription.getDefaultPlan()
+      : _subscription.getPlan(_membership.planId);
+
+    uint256 _totalPrice = _plan.price * _amountOfTerms;
+
+    if (_totalPrice / _amountOfTerms != _plan.price) {
```

Comment:
i feel like this could be manipulated somehow with certain values of `_plan.price` to get it to wrap in on itself and get an incredibly large number of terms

I feel like we should cap `_amountOfTerms` to some reasonable number instead (ex. up to 24 terms, 2 years probably)

Reply:
The plans already have a `maxTerms` value, so we are doing this check on subscription.